### ISSUE TITLE
Exploiting "abstract" + incorrect "Proof using" allows to abuse the discharging mechanism

### ISSUE THREAD

**Issue by Matafou** [coq/coq#19861]
### Description of the problem

Hi.

Sorry if this is an already known proof of false. I didn't find any reference to it.

By using `abstract` we abuse the check of `Proof using` at Qed (this is the bug imho), which in turn abuses the section discharging mechanism. Leading to a wrongly type lemma and an easy proof of false outside the section. 

I detected this because `Print_Assumption` and `coqchk` both refused to accept the resulting `.vo` file, which is comforting. I am surprised that this was never triggered though: I tested several versions between 8.20 down to 8.13 and all versions seem equally affected.

More details:

- the bug is triggered with `Proof using.` or equivalents like  `Proof using Type.`
- it is not triggered with "Proof." or "Proof using HypFalse."

This latter point suggests that the mechanism to "see" through the abstract exists, but maybe it is not correctly called when the `Proof using` is given by the user?

### Small Rocq / Coq file to reproduce the bug

```coq
Section S.
  Hypothesis HypFalse : False.
  Lemma false_in_section : False.
  Proof using.
    abstract (apply HypFalse).
  Qed.
End S.

Goal False.
Proof.
  apply false_in_section.
Qed.
```

### Version of Rocq / Coq where this bug occurs

8.13
8.19
8.20

**Comment by ppedrot** [comment]
Whoopsie, looks serious indeed and new AFAIK.

**Comment by ppedrot** [comment]
That said, `false_in_section` rightfully fails on master with
```
The following section variable is used but not declared:
HypFalse.

You can either update your proof to not depend on HypFalse, or you can update your Proof line from
Proof using 
to
Proof using HypFalse
```
Was this fixed consciously?

**Comment by Matafou** [comment]
Good news! 
Again it is surprising that it didn't show up earlier because `abstract` appears in hints (isn't this prohibited btw?):

```
Require  Import ZArith.
Lemma nat_lem (n:nat): n=0.
Proof.
  Print Hint.
(* In the database zarith:
    --> (*external*) (abstract Lia.lia)(level 10, pattern _ = _, id 0) *)
```

So there are many ways to trigger this. I was bitten by this via `intuition` that calls `auto with *`.

**Comment by Matafou** [comment]
Maybe people are just careful when they write `using` and I am a careless user :-).

**Comment by JasonGross** [comment]
First bad version is 8.11, according to
```bash
do docker run --platform linux/amd64 -it coqorg/coq:8.$i bash -c 'echo "Section S.  Hypothesis HypFalse : False.  Lemma false_in_section : False.  Proof using.    abstract (apply HypFalse).  Qed. End S. Goal False. Proof.   apply false_in_section. Qed." > foo.v; printf "%s %s\n" "$(coqc --print-version | cut -f1 -d" " )" "$(if coqc -q foo.v 2>/dev/null; then printf "passes"; else printf "fails"; fi)"'; done
```

**Comment by JasonGross** [comment]
I guess this just needs a dev/doc/critical entry?

**Comment by ppedrot** [comment]
> I guess this just needs a dev/doc/critical entry?

And a test.

> Was this fixed consciously?

Answering to myself about the PR I wrote: neither consciously nor unconsciously. I remember observing at the time that the environment we were using was wrong but I brushed it off as something that didn't matter in practice. I was wrong, but at least the gut feeling about the way things should look like indirectly made me solve the issue.
